### DISCUSSION TITLE
What is the case of  "installed locally" in `Bundler::Source::Rubygems#to_s`?

### DISCUSSION THREAD

**Discussion by junaruga** [ruby/rubygems#5352] [Q&A] [answered]
This question is related to this PR https://github.com/rubygems/rubygems/pull/5029 .

Seeing the latest master branch

https://github.com/rubygems/rubygems/blob/4b0d970cd31a7a34016bb6668f9756466070d6aa/bundler/lib/bundler/source/rubygems.rb#L107

I see that the method `Bundler::Source::Rubygems#to_s` is used to print the string expression of a Gem repository's source.

For the above PR, the following kind of message is printed.

```
+ bundle install
Your Gemfile contains scoped sources that don't implement a dependency API, namely:

  * rubygems repository http://127.0.0.1:8802/ or installed locally
```

But what is actually the case of "or installed locally"? In the source code, the case is `@allow_remote && @allow_local`, both flags are true. Can the `bundle install` install a gem from not only remote repository source but also from gems stored from local? What is the actual `bundle` command to demonstrate the case "or installed locally"?

Thanks.

**Comment by deivid-rodriguez** [comment]
If a gem is already installed locally, `bundle install` will will usually skip installing it. I think that's what "or installed locally" means. Sorry I haven't yet reviewed your PR, by the way.

**Reply by junaruga** [reply]
Thanks for the explanation. I understand it.

>  Sorry I haven't yet reviewed your PR, by the way.

Yeah, I am looking forward to see the PR will be reviewed and merged :)

**Reply by junaruga** [reply]
Reading the message again, "or installed locally" means "rubygems repository http://127.0.0.1:8802/ or (a gem belonging on the rubygem repository is already) installed locally", right?

The actual case for the "installed locally" is like this? (Edited)

```
$ cat Gemfile
source 'https://rubygems.org'

gem 'gem2rpm'
gem 'byebug'
```

Run "bundle install". Installed gems "gem2rpm" and "byebug".

```
$ bundle install
```

Add "rspec".

```
$ vi Gemfile

$ cat Gemfile
source 'https://rubygems.org'

gem 'gem2rpm'
gem 'byebug'
gem 'rspec'
```

It seems gem "byebug" and "gem2rpm" are skipped with "Using byebug 11.1.3" and "Using gem2rpm 1.0.2"

```
$ bundle install
Fetching gem metadata from https://rubygems.org/........
Resolving dependencies...
Using bundler 1.16.1
Using byebug 11.1.3
Fetching diff-lcs 1.5.0
Installing diff-lcs 1.5.0
Using gem2rpm 1.0.2
Fetching rspec-support 3.11.0
Installing rspec-support 3.11.0
Fetching rspec-core 3.11.0
Installing rspec-core 3.11.0
Fetching rspec-expectations 3.11.0
Installing rspec-expectations 3.11.0
Fetching rspec-mocks 3.11.0
Installing rspec-mocks 3.11.0
Fetching rspec 3.11.0
Installing rspec 3.11.0
Bundle complete! 3 Gemfile dependencies, 9 gems now installed.
Bundled gems are installed into `./app`
```

**Comment by deivid-rodriguez** [comment, marked as answer]
Yes!

**Reply by junaruga** [reply]
Thanks!
